### ISSUE TITLE
typing.TYPE_CHECKING doesn't seem to be honored

### ISSUE THREAD

**Issue by fayazkhan** [davidhalter/jedi#1127]
`typing.TYPE_CHECKING` is useful to avoid cyclic dependencies but still have type hinting. But it doesn't seem to be honored at all in jedi.

**Comment by davidhalter** [comment]
So what would happen if we honored it? I don't see the difference, Jedi doesn't have issues with cyclic dependencies anyway.

**Comment by fayazkhan** [comment]
I may have explained this poorly.

https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

From the docs:

> A special constant that is assumed to be True by 3rd party static type checkers. It is False at runtime.  Usage:
>
> ```
> if TYPE_CHECKING:
>     import expensive_mod
> 
> def fun(arg: 'expensive_mod.SomeType') -> None:
>     local_var: expensive_mod.AnotherType = other_fun()
>```
> 
> Note that the first type annotation must be enclosed in quotes, making it a “forward reference”, to hide the expensive_mod reference from the interpreter runtime. Type annotations for local variables are not evaluated, so the second annotation does not need to be enclosed in quotes.

So the idea is to solve cyclic dependencies with annotations.
(Wasn't talking about Jedi itself.)

**Comment by davidhalter** [comment]
I'm a bit confused now. I certainly understand what this variable is supposed to be doing, but I'm not sure if we need to change something in Jedi because of it.

Do you have an example where something is not working in Jedi because of it?

**Comment by njsmith** [comment]
We recently ran into this as well, in the [trio](https://github.com/python-trio/trio) project.

Our `trio.socket` module is a more-or-less drop-in replacement for the stdlib `socket` module, so we want to re-export all the many many constants that the socket module defines: `AF_INET`, `AF_INET6`, etc. etc. Originally we did this with a simple dynamic loop at runtime that simply copies all the UPPERCASE entries from `socket.__dict__` into `trio.socket.__dict__`, but of course no static analysis tool can make any sense of that, and we've been getting complaints about things like bad code completion in IDEs. But, we can't easily make the list static either, because the exact set of constants that the `socket` module exports depends on Python version, OS, the age of the system Python was built on, etc. etc. So after a [long discussion](https://github.com/python-trio/trio/issues/699), we decided to do something like:

```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from socket import (
        # long list of all the socket module constants we've ever heard of,
        # for the benefit of jedi/pylint/mypy/pycharm/...
        AF_INET, AF_INET6, ...
    )

# The actual re-exports we use at runtime
globals().update({name: value for (name, value) in socket.__dict__.items() if name.isupper()})
```

But... it turns out this doesn't work with jedi, because jedi is clever enough to realize that `TYPE_CHECKING` is `False`, so it still says that `trio.socket.AF_<TAB>` has no completions!

It looks like fixing this in jedi would be pretty simple? We could add a check to `jedi.evaluate.syntax_tree.eval_atom`, inside the `if atom.type == 'name'` block, that basically does `if atom.value == "TYPE_CHECKING": return True`.

CC: @jmfrank63

**Comment by njsmith** [comment]
For the record, it looks like mypy's version of the check is hard-coded to assume that any variable named `TYPE_CHECKING` or attribute lookup like `<anything>.TYPE_CHECKING` is always `True`: see [here](https://github.com/python/mypy/blob/4acd48e5cd2a38ac9f6b57dcd9a41b0551cf53a6/mypy/semanal.py#L3796-L3799) and then a few lines down [here](https://github.com/python/mypy/blob/4acd48e5cd2a38ac9f6b57dcd9a41b0551cf53a6/mypy/semanal.py#L3820-L3821).

**Comment by davidhalter** [comment]
This won't be fixed until the typeshed (#1096) branch is merged. It's already working in there, but a lot of other things aren't.

**Comment by davidhalter** [comment]
This branch was merged. TYPE_CHECKING should be working as intended now. Please test!
